sdk/src/main/java/com/mapbox/maps/ViewAnnotationManagerImpl.kt

```diff
@@ -308,31 +316,54 @@ internal class ViewAnnotationManagerImpl(
   private fun positionAnnotationViews(
-    positionDescriptorCoreList: List<ViewAnnotationPositionDescriptor>
+    positionDescriptorUpdatedList: List<ViewAnnotationPositionDescriptor>
   ) {
+    val needToReorderZ = needToReorderZ(
+      positionDescriptorCurrentList,
+      positionDescriptorUpdatedList,
+    )
+    positionDescriptorCurrentList = positionDescriptorUpdatedList
+
     // as per current implementation when view annotation was added with WRAP_CONTENT dimension AND
     // allowOverlap = false - core will notify only about this particular view and thus we will remove
     // all the others and then restore later resulting in a quick blink
-    val wrapContentWithAllowOverlapFalseViewAdded = positionDescriptorCoreList.size == 1 &&
-      (positionDescriptorCoreList[0].width == WRAP_CONTENT || positionDescriptorCoreList[0].height == WRAP_CONTENT)
+
+    // FIXME - other bugs are possible, this does not cover all cases, e.g. 3 annotations and the last one blinks :
+    // Positions : 3
+    //   pos : [identifier: 44, width: 282, height: 196, leftTopCoordinate: [x: 148.00000000000566, y: 616.9999999999659]]
+    //   pos : [identifier: 45, width: 282, height: 196, leftTopCoordinate: [x: 483.0000000000149, y: 566.9999999999759]]
+    //   pos : [identifier: 48, width: -2, height: -2, leftTopCoordinate: [x: -2.147483218E9, y: -4.2949667920000005E9]]
+    // Positions : 3
+    //   pos : [identifier: 42, width: 282, height: 196, leftTopCoordinate: [x: 99.99999999998583, y: 368.9999999999887]]
+    //   pos : [identifier: 44, width: 282, height: 196, leftTopCoordinate: [x: 148.00000000000566, y: 616.9999999999659]]
+    //   pos : [identifier: 45, width: 282, height: 196, leftTopCoordinate: [x: 483.0000000000149, y: 566.9999999999759]]
+    // Positions : 1
+    //   pos : [identifier: 49, width: -2, height: -2, leftTopCoordinate: [x: -2.147483362E9, y: -4.2949661230000005E9]]
+    // Positions : 3
+    //   pos : [identifier: 42, width: 282, height: 196, leftTopCoordinate: [x: 99.99999999998583, y: 368.9999999999887]]
+    //   pos : [identifier: 44, width: 282, height: 196, leftTopCoordinate: [x: 148.00000000000566, y: 616.9999999999659]]
+    //   pos : [identifier: 45, width: 282, height: 196, leftTopCoordinate: [x: 483.0000000000149, y: 566.9999999999759]]
+    val wrapContentWithAllowOverlapFalseViewAdded = positionDescriptorUpdatedList.size == 1 &&
+      (positionDescriptorUpdatedList[0].width == WRAP_CONTENT || positionDescriptorUpdatedList[0].height == WRAP_CONTENT)
     // firstly delete views that do not belong to the viewport if it's not specific use-case from above
     if (!wrapContentWithAllowOverlapFalseViewAdded) {
-      currentViewsDrawnMap.keys.forEach { id ->
-        if (positionDescriptorCoreList.indexOfFirst { it.identifier == id } == -1) {
+      currentlyDrawnViewIdSet.forEach { id ->
+        if (positionDescriptorUpdatedList.indexOfFirst { it.identifier == id } == -1) {
```

Comment:
if I understand correctly in worst case it can be O(N)? 
maybe save somehow id + position?